src/sentry/debug_files/artifact_bundles.py

```diff
@@ -0,0 +1,114 @@
+from typing import Mapping, Sequence
+
+from sentry.models.artifactbundle import ArtifactBundle, ArtifactBundleArchive
+from sentry.utils import metrics
+
+
+# TODO:
+class ArtifactUrlIndex:
+    pass
+
+
+def index_artifact_bundles_for_release(
+    artifact_bundles: Sequence[ArtifactBundle], release: str = "", dist: str = ""
+):
+    """
+    This indexes the contents of `artifact_bundles` into the database, using
+    the given `release` and `dist` pair.
+
+    Synchronization is achieved using the 4-state `indexing_state`.
+    It is flipped from `needs_indexing` to `is_indexing` to avoid conflicts in
+    case two different jobs are calling this method at the same time with the
+    same `artifact_bundle`.
+    Once indexing is done, the state is set to `was_indexed`.
+
+    To avoid too much load and queries on the database, the provided
+    `artifact_bundles` are first merged/deduplicated into an in-memory structure.
+
+    Then, indexing of individual files happens in the following way:
+    - If a key already exists with a higher `date_last_modified`, it is ignored
+      as that indicates out-of-order calls of this method. We always want the
+      most up-to-date bundle to win.
+    - Otherwise, the key is created or updated to point to the `artifact_bundle`.
+    """
+
+    # First: Mark the bundles as `is_indexing` using a compare-and-swap like update
+    started_artifact_bundles = []
+    for artifact_bundle in artifact_bundles:
+        did_start_indexing = ArtifactBundle.objects.filter(
+            id=artifact_bundle.id, indexing_state="needs_indexing"
+        ).update(indexing_state="is_indexing")
+        if did_start_indexing:
+            started_artifact_bundles.append(artifact_bundle)
+        else:
+            # a different job is already taking care of this bundle
+            metrics.incr("artifact_bundle_indexing.conflict")
+    artifact_bundles = started_artifact_bundles
+
+    if not artifact_bundles:
+        return
+
+    # Then: Merge the bundles into an in-memory index
+    files_to_index: Mapping[str, ArtifactBundle] = {}
+
+    for artifact_bundle in artifact_bundles:
+        archive = ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=False)
+        try:
+            for info in archive.manifest.get("files", {}).values():
+                url = info.get("url")
+                if not url:
+                    continue
+
+                indexed = files_to_index.get(url)
+                if not indexed or indexed.date_last_modified < artifact_bundle.date_last_modified:
+                    files_to_index[url] = artifact_bundle
+
+        finally:
+            archive.close()
+
+    # this is used for partitioning
+    date_touched = datetime.now()
+
+    # And finally, commit the intended changes to the database, carefully merging
+    # with the state that already exists in the database.
+    indexed_urls = 0
+    for url, artifact_bundle in files_to_index:
```

Comment:
In my view of the implementation, now given the `files_to_index` I would group all the urls by bundle_id and start a transaction to update all the urls of each bundle_id group. This is to have medium granularity of updates while still keeping transactional guarantees.

```python
bundle_id_urls = defaultdict(set)
for url, artifact_bundle in files_to_index.items():
	bundle_id_urls[artifact_bundle].add(url)
```